docs/proposals/24-dynamic-node-groups-for-shoot-cluster-autoscaler.md

```diff
@@ -0,0 +1,338 @@
+
+
+The above will permit the MCM CloudProvider to discover all the Machine Deployments for the shoot cluster.
+
+### 3. Dynamic Node Group Sizing.
+
+The [cloudprovider.CloudProvider](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L100) interface is the facade between the CA and the cloud platform.
+
+[cloudprovider.NodeGroup](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L163) is the CA abstraction used to represent the set of cloud-provider nodes which are in-turn represented by [cloudprovider.Instances](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/cloud_provider.go#L238)
+
+The CA carries out scale-down and scale-up operations within a `NodeGroups` respecting the constraints given by the `NodeGroup.MinSize()` and `NodeGroup.MaxSize()` methods.
+
+The CA runs a reconcile loop every `scanPeriod` interval (default: `10s`). At the beginningo of the reconcile loop, the CA invokes `CloudProvider.Refresh()` to permit the cloud provider implementation to update its cache and then issues a call to `CloudProvider.NodeGroups()` to retrieve the latest `[]NodeGroup`.  The MCM CloudProvider implements this by interrogating the machine deployments.
+Each `NodeGroup` continues to be associated with its corresponding `MachineDeployment` for the zone.
+
+We will offer 2 strategies for node-group sizing: *Lax Greedy Sizing* and *Equitable Sizing*
+
+### 3.1 Lax-Greedy Sizing
+1. If this strategy is selected, then for each `NodeGroup` the `NodeGroup.MaxSize()` is initially returned as the `MachineDeployment.Spec.PoolMaximum` and the `NodeGroup.MinSize()` is given as the `0` if there are no nodes provisioned for the shoot cluster.
+1. The CA will execute scale-up/scale-down activities for the node groups if needed. If a NodeGroup does not have quota it will back-off and try another NodeGroup.
+1. For the next scane, we compute `NodeGroup.MaxSize()` as follows
+     1. `NodeGroup.MaxSize()=      PoolMaximum-CountOfAllNodesMaterializedInOtherNodeGroups)`
+
+
+#### Pros/Cons
+1. Pro: This strategy has the primary advantage that workload can be provisioned within the zone for the NodeGroup. It can be considered as work-load friendly
+2. Pro: It can handle back-off well. If nodes can't be provisioned in one node group, the other node group can take up the slack.
+3. Con: It isn't backward compatabile with our current distribution logic.
+4. Con: We need to add support for scale from zero for extended/ephmeral resources. We have [Gardener CA #132](https://github.com/gardener/autoscaler/issues/132) for this.
+
```

Comment:
Another downside of this is that it may not use zones at all, i.e. all instances may stay in one availability zone even if multiple are configured. While this may be advantageous short-term, it may impact availability if the zone goes down. The potential impact in terms of availability/balance needs to be clear to users.

Reply:
adding as Con

Reply:
>Another downside of this is that it may not use zones at all, i.e. all instances may stay in one availability zone even if multiple are configured.  

Ideally this should not be the concern of CA on which node group it should scale up. Instead the onus should be on the consumer to appropriately set TSC or alternative policies to force distribution of pods across zones which in turn will force CA to choose another zone, if a new node in another zone cannot be launched then the pod will remain in pending state.

Reply:
@himanshu-kun pointed that CA also provides [balance-similar-node-groups](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/main.go#L181) CLI flag which allows CA to make best effort to balance launching of nodes amongst similar node groups. AFAIK this would also help distribute the nodes amongst similar node groups and should continue to work even with the `Lax` strategy. However we should probably test this.